### ISSUE TITLE
Cable Trace View Readability

### ISSUE THREAD

**Issue by mshuttle76** [netbox-community/netbox#8450]
### NetBox version

v3.1.6

### Feature type

Change to existing functionality

### Proposed functionality

![image](https://user-images.githubusercontent.com/60605051/150965989-8915b7c8-3150-4f6a-9c57-677febc7cf7e.png)

For long object coordinates (site, rack group, rack ecc.) it would be useful to be rendered in two lines instead of a single one for more readability.

### Use case

Improved readability for objects into cable trace view.

**Comment by jeremystretch** [comment]
IIRC the root issue here is that SVG has no notion of text wrapping; it needs to be handled manually, which is not trivial. I could be wrong though.

**Comment by DetunizedGravity** [comment]
I would like to raise this issue from the (soon to be) dead: we were forced to rename all our sites to very short versions of themselves, putting the actual names in the description field, in order to keep the cable trace readable. That is a less than ideal situation.

**Comment by jeremystretch** [comment]
@DetunizedGravity are you interested in owning the issue?

**Comment by DetunizedGravity** [comment]
To sum up what I answered by email (and that I thought would appear here), I don't think I have enough free time to commit to anything, and I have stopped coding seriously for so long that I'm completely lost in the current devops ecosystem.

I just tried to see if I could install Visual Studio Code and evaluate the actual cost of entry and... Oh boy how steep the learning curve seems to be.

So, definitely no at the moment.

**Comment by dreng** [comment]
Personally I've never run into that problem but the example above shows a very annoying bug.

What about just appending every variable with labels.append instead of concatenating some of them in one line (in svg.py, classmethod _get_labels)? That would be a very quick fix. It leads to the situation that you might end up with a huge amount of lines (it sums up to 5 lines at most in our environment), but that would be a much better solution than leaving it as it is.

I just fixed this on my test server and could add a PR if you consider this as an acceptable solution.

**Comment by jeremystretch** [comment]
@dreng that might help but I'm not sure it would totally resolve the issue for very long strings. Either way, I'm about to make some major changes to the cable trace rendering logic for #9102, so it's probably best to hold off for now.

I'm going to mark this as blocked by #9102 and revisit during/after that work.

**Comment by jeremystretch** [comment]
The labeling still works pretty much the same after #9102 (although the code has been cleaned up significantly). Leaving this open for anyone who would like to take a stab at fitting the text into the SVG more elegantly.

**Comment by dreng** [comment]
@jeremystretch I doubt that there's an elegant solution we can work out at compile time. There are several problems that prevent that. First of all we use a font-family. You can not expect any browser on any OS to render the font identically. Second of all we use proportional fonts which makes it more than difficult to calculate the width of a certain string. E.g. 20 times "i" takes the space of 70px on my system while 20 times "M" takes 206px (about 3 times as much!).

I think the only true solution would be to determine the width of the string at runtime and add the needed padding. That could probably be achieved using Javascript but I have no idea how to do that.

To sum it up we need a good compromise and should accept that there might be not an ideal solution.

**Comment by arthanson** [comment]
Potential solution - 
[textflow.zip](https://github.com/netbox-community/netbox/files/9570504/textflow.zip)
There is a JS implementation here (https://old.carto.net/papers/svg/textFlow/) (haven't tried it yet), files attached as the website looked like it sometimes become non-accessible.

**Comment by DetunizedGravity** [comment]
Hello Jeremy,

I am sorry but it would be a bad idea. I haven't touched coding for so long
that git did not exist yet the last time I pushed code into the open (Did
you know there was another "pysocks" project on sourceforge before the the
one you surely know now was born? :) ). I'd have to (re)learn how to use
all the modern tooling before I could play an active part. Not to mention
modern python, django, docker, and the netbox code base. What leads me to
the second obstacle: I **really** do not have enough time on my hands for
that right now.

Netbox is already important to us and it will only become even more so with
time. So it would be nice if we could play a more active role in its
development. Unfortunately our current workload makes it impossible to most
if not all of us right now.

It may change in a few months from now, when things will have settled a bit
and I should be able to get my head out of the sand. I hope.

Best regards,
  Xavier Lagraula


Le mer. 30 mars 2022 à 18:42, Jeremy Stretch ***@***.***> a
écrit :
